default.nix

```diff
@@ -0,0 +1,179 @@
+  untarAndWrap = name: cmds: ''
+    mkdir -p $out/bin
+    tar xzvf ./${name}.tgz -C $out --strip-components=1
+    if [ "$installJavascript" = "1" ]; then
+      cp -R node_modules $out/
+      ${ concatStringsSep ";" (map (cmd:
+        ''makeWrapper ${cmd} $out/bin/${baseNameOf cmd} --run "cd $out"''
+      ) cmds) }
+    fi
+  '';
+in {
+  buildNpmPackage = args @ {
+    src, npmBuild ? "npm ci", npmBuildMore ? "",
+    buildInputs ? [], npmFlags ? [], ...
+  }:
+    let
+      info  = npmInfo src;
+      lock  = fromJSON (readFile (src + "/package-lock.json"));
+    in stdenv.mkDerivation ({
+      inherit (info) name;
+      inherit (info.info) version;
+
+      preBuildPhases    = [ "npmCachePhase" ];
+      preInstallPhases  = [ "npmPackPhase" ];
+
+      npmCachePhase = ''
+        addToSearchPath NODE_PATH ${npmModules}   # pacote
+        node ${./mknpmcache.js} ${cacheInput "npm-cache-input.json" lock}
+      '';
+
+      buildPhase = ''
+        ${npmAlias}
+        runHook preBuild
+        ${npmBuild}
+        ${npmBuildMore}
+        runHook postBuild
+      '';
+
+      # make a package .tgz (no way around it)
+      npmPackPhase = ''
+        ${npmAlias}
+        npm prune --production
+        npm pack --ignore-scripts
+      '';
+
+      installPhase = untarAndWrap info.name [npmCmd];
+    } // commonEnv // args // {
+      buildInputs = commonBuildInputs ++ buildInputs;
+      npmFlags    = npmFlagsNpm ++ npmFlags;
+    });
+
+  buildYarnPackage = args @ {
+    src, yarnBuild ? "yarn", yarnBuildMore ? "", integreties ? {},
+    buildInputs ? [], yarnFlags ? [], npmFlags ? [], ...
+  }:
+    let
+      info        = npmInfo src;
+      deps        = { dependencies = fromJSON (readFile yarnJson); };
+      yarnIntFile = writeText "integreties.json" (toJSON integreties);
+      yarnLock    = src + "/yarn.lock";
+      yarnJson    = runCommand "yarn.json" {} ''
+        set -e
+        addToSearchPath NODE_PATH ${npmModules}             # ssri
+        addToSearchPath NODE_PATH ${yarn2nix.node_modules}  # @yarnpkg/lockfile
+        ${_nodejs}/bin/node ${./mkyarnjson.js} ${yarnLock} ${yarnIntFile} > $out
+      '';
+    in stdenv.mkDerivation ({
+      inherit (info) name;
+      inherit (info.info) version;
+
+      preBuildPhases    = [ "yarnConfigPhase" "yarnCachePhase" ];
+      preInstallPhases  = [ "yarnPackPhase" ];
+
+      # TODO
+      yarnConfigPhase = ''
+        cat <<-END >> .yarnrc
+
+        	yarn-offline-mirror "$PWD/yarn-cache"
+        	script-shell "${shellWrap}/bin/npm-shell-wrap.sh"
+        	nodedir "${_nodejs}"
+        END
+      '';
+
+      yarnCachePhase = ''
+        mkdir -p yarn-cache
+        node ${./mkyarncache.js} ${cacheInput "yarn-cache-input.json" deps}
+      '';
+
+      buildPhase = ''
+        ${npmAlias}
+        ${yarnAlias}
+        runHook preBuild
+        ${yarnBuild}
+        ${yarnBuildMore}
+        runHook postBuild
+      '';
+
+      # TODO: install --production?
```

Comment:
I think package makers really like using custom commands instead of adding a pack command. (like vault using just `build`). Maybe there should be a way to add these.

Reply:
I'm currently using `yarnBuildMore` for that. See [the vault gist](https://gist.github.com/obfusk/4abe41103de8c3a46fb0d2352bcf9e72). Did you have something else in mind?